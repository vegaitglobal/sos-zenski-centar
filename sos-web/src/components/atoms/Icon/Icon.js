import { ReactComponent as Logo } from './svg/logo.svg';

export const Icon = {
  Logo,
};
