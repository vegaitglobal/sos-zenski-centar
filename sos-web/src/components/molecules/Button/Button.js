import { StyledButton } from './Button.styles';

export const Button = (props) => {
  return <StyledButton {...props}>{props.children}</StyledButton>;
};
