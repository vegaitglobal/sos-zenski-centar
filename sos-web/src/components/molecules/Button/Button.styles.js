import styled from 'styled-components';
import { color } from '../../../styles/config/theme';
import { typeStyle } from '../../../styles/config/typeStyles';

export const StyledButton = styled.button`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 14px;
  min-width: 145px;
  max-width: 100%;
  ${typeStyle.medium};
  color: ${color.greyLightest};
  background-color: ${color.purple};
  border-radius: 8px;
  transition: opacity 0.2ms ease-in-out;

  &:disabled {
    opacity: 0.5;
  }
`;
