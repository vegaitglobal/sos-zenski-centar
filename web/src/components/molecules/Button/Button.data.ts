import { ReactNode } from 'react';
import { NativeProps } from '../../atoms/AbstractButton/AbstractButton.data';

export enum ButtonTheme {
	Dark = 'dark',
	Light = 'light',
}

export interface StyledButtonProps {
	$buttonTheme: ButtonTheme;
	$onlyIcon: boolean;
}

export interface ButtonProps extends NativeProps {
	icon?: ReactNode;
	buttonTheme?: ButtonTheme;
}
