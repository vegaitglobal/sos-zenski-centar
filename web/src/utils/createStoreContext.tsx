import { createContext, PropsWithChildren, FunctionComponent, Context, useContext } from 'react';
import useStoreFactory from '../hooks/useStoreFactory';

/**
 * Creates a context for the store created by the passed `storeFactory`.
 * Additionally returns the `Provider` component and the hook for the created store's context.
 *
 * @param name - name of the store
 * @param storeFactory - a function that returns an object representing the store
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const createStoreContext = <TStore extends Record<string, unknown>>(
	name: string,
	storeFactory: () => TStore,
): [FunctionComponent<unknown>, () => TStore, Context<TStore>] => {
	const context = createContext<TStore>({} as TStore); // the initial value is only to satisfy TS and won't be used

	/**
	 * Initializes and contains the context for the `TStore`.
	 */
	function StoreProvider({ children }: PropsWithChildren<unknown>): JSX.Element {
		const store = useStoreFactory(storeFactory);
		return <context.Provider value={store}>{children}</context.Provider>;
	}

	const useStore = (): TStore => {
		const store = useContext(context);
		if (store) {
			return store;
		}
		throw new ReferenceError(
			`${name} is not initialized, make sure that your component is wrapped in the 'Provider' of the store.`,
		);
	};

	return [StoreProvider, useStore, context];
};

export default createStoreContext;
