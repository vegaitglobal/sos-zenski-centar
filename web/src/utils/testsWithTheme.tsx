import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { ReactElement, ComponentType, ReactNode } from 'react';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType = ({ children }) => componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
