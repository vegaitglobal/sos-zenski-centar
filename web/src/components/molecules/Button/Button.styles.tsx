import { size } from 'polished';
import styled, { css, DefaultTheme, FlattenInterpolation, ThemeProps } from 'styled-components';
import { ease } from '../../../styles/config/easing';
import { typeStyle } from '../../../styles/config/typeStyles';
import { FontWeight, TransitionDuration } from '../../../styles/config/variables';
import { AbstractButton } from '../../atoms/AbstractButton/AbstractButton';
import { ButtonTheme, StyledButtonProps } from './Button.data';

const dark = css`
	background-color: ${({ theme }) => theme.color.black};
	color: ${({ theme }) => theme.color.white};

	&:hover:not(:disabled),
	&:focus:not(:disabled) {
		background-color: ${({ theme }) => theme.color.blackFaded};
	}
`;

const light = css`
	background-color: ${({ theme }) => theme.color.white};
	color: ${({ theme }) => theme.color.black};

	&:hover:not(:disabled),
	&:focus:not(:disabled) {
		background-color: ${({ theme }) => theme.color.whiteFaded};
	}
`;

const disabled = css`
	background-color: ${({ theme }) => theme.color.blackFaded};
	color: ${({ theme }) => theme.color.whiteFaded};
	cursor: not-allowed;
`;

const onlyIcon = css`
	${size('3.5rem')};
	min-height: auto;
	border-radius: 50%;
	padding: 0;

	svg {
		${size('1.6rem')};
		margin-left: 0;
	}
`;

const themes: { [key in ButtonTheme]: FlattenInterpolation<ThemeProps<DefaultTheme>> } = {
	dark,
	light,
};

export const StyledButton = styled(AbstractButton)<StyledButtonProps>`
	${typeStyle.normal};
	font-weight: ${FontWeight.Bold};
	display: inline-flex;
	align-items: center;
	justify-content: center;
	padding: 0.2rem 2.5rem;
	min-height: 4rem;
	border-radius: 4.6rem;
	transition: ${TransitionDuration.Normal} ${ease.easeOutSine};
	transition-property: color, background-color;

	${({ $buttonTheme }) => themes[$buttonTheme]};

	&:disabled {
		${disabled};
	}

	svg {
		${size('2.6rem')};
		margin-left: 1rem;
	}

	${({ $onlyIcon }) => $onlyIcon && onlyIcon}
`;
