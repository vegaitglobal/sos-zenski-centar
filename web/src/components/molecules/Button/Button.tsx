import { forwardRef, MouseEvent } from 'react';
import { ElementType } from '../../atoms/AbstractButton/AbstractButton.data';
import { ButtonProps, ButtonTheme } from './Button.data';
import { StyledButton } from './Button.styles';

export const Button = forwardRef<ElementType, ButtonProps>(
	({ children, icon, buttonTheme = ButtonTheme.Dark, onClick, ...props }, ref) => {
		const handleClick = (event: MouseEvent<ElementType>): void => {
			event.currentTarget.blur();
			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
			// @ts-ignore
			onClick?.(event);
		};

		return (
			<StyledButton
				{...props}
				ref={ref}
				$buttonTheme={buttonTheme}
				$onlyIcon={!children && !!icon}
				onClick={handleClick}
				data-testid="Button"
			>
				{children}
				{icon}
			</StyledButton>
		);
	},
);
