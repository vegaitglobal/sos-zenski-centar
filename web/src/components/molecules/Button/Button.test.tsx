import renderer from 'react-test-renderer';
import Button from './index';
import { renderWithTheme, componentWithTheme } from '../../../utils/testsWithTheme';
import { Icon } from '../../atoms/Icon/Icon';

describe('Button', () => {
	describe('Render', () => {
		it('Renders Button', () => {
			const { getByTestId } = renderWithTheme(<Button />);

			const element = getByTestId('Button');

			expect(element).toBeTruthy();
		});

		it('Renders Icon', () => {
			const { getByTestId } = renderWithTheme(<Button icon={<Icon.Logo />} />);

			const element = getByTestId('Button');
			const child = element.querySelector('svg');

			expect(element).toContainElement(child);
		});

		it('Can be anchor tag', () => {
			const { getByTestId } = renderWithTheme(<Button href="https://www.google.com" />);

			const element = getByTestId('Button');

			expect(element).toContainHTML('</a>');
			expect(element).toHaveAttribute('href', 'https://www.google.com');
		});

		it('Can be button tag', () => {
			const { getByTestId } = renderWithTheme(<Button />);

			const element = getByTestId('Button');

			expect(element).toContainHTML('</button>');
		});

		it('Renders correctly', () => {
			const tree = renderer.create(componentWithTheme(<Button />)).toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('Can have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Button>
					<span>Child</span>
				</Button>,
			);

			const element = getByTestId('Button');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});

	describe('Attributes', () => {
		it('Can be disabled', () => {
			const { getByTestId } = renderWithTheme(<Button disabled />);

			const element = getByTestId('Button');

			expect(element).toBeDisabled();
		});
	});
});
