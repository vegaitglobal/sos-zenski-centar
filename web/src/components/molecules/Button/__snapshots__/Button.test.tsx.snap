// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render Renders correctly 1`] = `
<button
  className="sc-bdnxRM hxtrhb"
  data-testid="Button"
  onClick={[Function]}
  type="button"
/>
`;
